src/librustc/ty/instance.rs

```diff
@@ -45,6 +45,51 @@ pub enum InstanceDef<'tcx> {
     CloneShim(DefId, Ty<'tcx>),
 }
 
+impl<'a, 'tcx> TyCtxt<'a, 'tcx, 'tcx> {
+    pub fn is_inline_instance(
+        self,
+        instance: &ty::Instance<'tcx>
+    ) -> bool {
+        use hir::map::DefPathData;
+        let def_id = match instance.def {
+            ty::InstanceDef::Item(def_id) => def_id,
+            ty::InstanceDef::DropGlue(_, Some(_)) => return false,
+            _ => return true
+        };
+        match self.def_key(def_id).disambiguated_data.data {
+            DefPathData::StructCtor |
+            DefPathData::EnumVariant(..) |
+            DefPathData::ClosureExpr => true,
+            _ => false
+        }
+    }
+
+    pub fn requests_inline(
```

Comment:
I think this should be called `requires_local_instance` or something else along those lines.